### PR TITLE
dosbox-staging: new port

### PR DESCRIPTION
**Pull request by kakuhen**
#### Description
Requested on IRC about a month ago by "Grounded0".

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.15.7 19H2026 x86_64
Xcode 12.4 12D4e

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: emulators/dosbox-staging/Portfile
Comment (reneeotten) [inline emulators/dosbox-staging/Portfile:7]: please add `github.tarball_from archives` and recalculate the `checksums`

File: emulators/dosbox-staging/Portfile
Comment (reneeotten) [inline emulators/dosbox-staging/Portfile:13]: this is the default since MacPorts 2.7 so the `platforms` line can/should be removed

File: emulators/dosbox-staging/Portfile
Comment (reneeotten) [inline emulators/dosbox-staging/Portfile:12]: please add your (obfuscated) e-mail as well

File: emulators/dosbox-staging/Portfile
Comment (reneeotten) [inline emulators/dosbox-staging/Portfile:37]: please align all dependencies the same as on the first line

**Review by reneeotten — CHANGES_REQUESTED** [review]
@kakuhen thanks for the submission. I haven't checked the logs for anything, but just a few comments based on a read through of the Portfile

**Comment by Grounded0** [comment]
As a caretaker of DOSBox Staging we would recommend to hold off release for now as we're days away from releasing 0.79.0. Better to get this project directly on current release. We're currently busy preparing that release and will be happy to help with preparing this port after its released.

File: emulators/dosbox-staging/Portfile
Comment (kakuhen) [reply emulators/dosbox-staging/Portfile:7]: What is the difference between archives and using source code from releases?

File: emulators/dosbox-staging/Portfile
Comment (kakuhen) [reply emulators/dosbox-staging/Portfile:12]: I am in the process of changing e-mails. The one I am currently using for commits will change soon. Once I have finished changing e-mails, I will update all of the ports I maintain to include my e-mail in the obfuscated format.

**Comment by Grounded0** [comment]
It's shipping:
https://dosbox-staging.github.io/v0-79-0/

**Comment by Grounded0** [comment]
I can operate as an intermediary between our people and your people so DOSBox Staging 0.79.0 can be delivered to MacPorts in grace.

**Comment by kakuhen** [comment]
Dependency was silently introduced... Thankfully it's not too hard to build. Just made a PR for it.

Assume no further work on this PR until iir1 is merged into the ports tree (dosbox-staging 0.79 does not build without it). And if lack of an e-mail is a problem, then I am willing to mark someone else here as the maintainer of the port, or `nomaintainer`.

**Comment by Grounded0** [comment]
Iir1 can be built statically via Meson wrap.

**Comment by kakuhen** [comment]
OK this update is proving a lot harder to incorporate. For some reason dosbox-staging wants to download its own speexdsp 1.2.1 even if the system already provides it. Configure fails with the following message.
```
meson.build:397:4: ERROR: Automatic wrap-based subproject downloading is disabled
```

I traced the changes to `meson.build` back to dosbox-staging commit 5111f8. Unfortunately, `meson-log.txt` does not record the results of running these tests, and I am not confident in build systems that silently download blobs without any form of integrity verification (e.g. checksums and GPG signatures). For instance, this is what happens when we do allow meson to download at will.

```
Downloading speexdsp source from https://downloads.xiph.org/releases/speex/speexdsp-1.2.1.tar.gz
Download size: 922584
Downloading: ..........
Downloading speexdsp patch from https://wrapdb.mesonbuild.com/v2/speexdsp_1.2.1-4/get_patch
Download size: 6224
```

There appears to be a file size check, at least, but there's various concerns in this approach that leads me to force ` --wrap-mode=nodownload`:

1. No indication of any checksum or GPG signature verification on these blobs
2. No indication of version control on these blobs, effectively killing the chance of reproducible builds.
3. I have no idea what happens when Meson attempts installing these files. My best guess is that these files will be outside of MacPorts' control, builds in trace mode may fail as a result, and we likely break any existing installation of the speexdsp port.

Conclusion: While iir1 has been added to MacPorts, I do not understand the FP issue with speexdsp (or what the patches do). So again, work on this port will stall until the speexdsp maintainer can do something about this issue.

**Comment by shermp** [comment]
@kakuhen look in `subprojects/speexdsp.wrap`, it has the source files, version and hashes that meson downloads. Once downloaded, the dependency is compiled by meson, and statically linked to the dosbox-staging executable. It is never installed anywhere outside of the project directory.

All of the Meson wraps behave this way. See https://mesonbuild.com/Wrap-dependency-system-manual.html for more info on Meson wraps.

I hope this provides some clarification.

**Comment by kcgen** [comment]
Hi @kakuhen, we do have a section for maintainers in the build docs, which I think can help here.

In the case of SpeexDSP, see the first para here (and speexdsp is used in the example setup, to force it to be pulled from the system):

https://github.com/dosbox-staging/dosbox-staging/blob/release/0.79.x/BUILD.md#repository-maintainer--packager-builds

But we also saw another maintainer wrestling with this, and so a recent commit has just been added that will by-default try the use the system-provided speexdsp; however, it will test it for floating-point API behavior, and if it fails it will fallback to the wrap - which we can control.

This commit will be backported to 0.79.1:

https://github.com/dosbox-staging/dosbox-staging/pull/1963

So at that point, you won't need `meson setup -Dsystem_libraries=speexdsp build`

Hope this helps, and thanks for all the work getting DOSBox Staging into macports!

**Comment by Grounded0** [comment]
DOSBox Staging 0.79.1 will be released in few days which incorporates enhancements that make it easier to package so wait for that before going ahead.

**Comment by kakuhen** [comment]
Thanks for the clarification, everybody! 

Sorry for the late response. Anyway, I just noticed that there is an update to the meson.build to test the SpeexDSP libraries. Hopefully what we have currently in ports works. I'm going to follow @Grounded0's advice and wait for the 0.79.1 release and see how things go from there.

**Comment by Grounded0** [comment]
DOSBox Staging 0.79.1 Maintenance Release Notes:
https://dosbox-staging.github.io/v0-79-1/

**Comment by kakuhen** [comment]
@Grounded0 Thanks for the notification. Sorry it took me a while to act on this information. I have successful builds on my end. Assuming the CI builds pass, we should be good to go for a final review and dosbox-staging will be in the ports tree!

I'd also like to thank the developers for being very receptive and working hard on new releases.

**Comment by kakuhen** [comment]
OK. CI builds are passing. Requesting another review.

**Comment by kakuhen** [comment]
@reneeotten ping

**Review by neverpanic — APPROVED** [review]

**Comment by kcgen** [comment]
@kakuhen , 

It would be nice to get it building on Mojave and Snow Leopard:



---

Here's a summary of the errors:

### Mojave

```
../dosbox-staging-0.79.1/src/hardware/lpt_dac.cpp:46:11:
error: no member named 'merge' in 'std::__1::set<ChannelFeature,
std::__1::less<ChannelFeature>, std::__1::allocator<ChannelFeature> >'
        features.merge(extra_features);
        ~~~~~~~~ ^
1 error generated.
```

Mjoave's clang-1001.0.46.4 wasn't C++17 compliant, given it reports that the `std::set` container doesn't have a `merge(...)` member function, when in fact [it  does](https://en.cppreference.com/w/cpp/container/set/merge). 

This can be worked-around without harming the code quality.

### Snow Leopard

```
../dosbox-staging-0.79.1/src/misc/cross.cpp:517:10: error: use of undeclared identifier 'strnlen'
                       strnlen(find_ext,
```

It appears `strnlen` was not available in the system headers under Snow Leopard.

This can also be worked-around.

**Comment by kcgen** [comment]
@kakuhen , is it possible to feed my `**[kc/clang-cleanups-1](https://github.com/dosbox-staging/dosbox-staging/tree/kc/clang-cleanups-1)** branch through these build bots, and check if it takes care of it?

**Comment by kakuhen** [comment]
Please use Trac to discuss issues. Unfortunately, I don't have time at the moment to investigate build failures for 10.6 or 10.14. Attach a main.log and any other details you may find important to a ticket on Trac. Thanks

**Comment by kcgen** [comment]
>  Unfortunately, I don't have time at the moment to investigate build failures for 10.6 or 10.14. 

I've investigated (and potentially fixed) these build failures, but I don't know how to test the proposed changes in macport's buildbot system.  

I will check on Trac; thanks for the tip, @kakuhen :+1:

**Comment by kencu** [comment]
the 10.6 failure is likely fixed by adding this near the top of the file:
```
PortGroup legacysupport 1.1
```

The 10.14 failure might be fixed by forcing a newer compiler:

```
PortGroup compiler_blacklist_versions 1.0
```

and then later on in the Portfile:
```
compiler.blacklist-append {clang < 1002}
```

**Comment by kcgen** [comment]
Thanks @kencu! On the source-side, I've merged changes to mitigate/work-around these, however I can't know for certain if they will. Hopefully also adding your suggestion on the port-side will further the odds that DOSBox Staging builds across the board, come the next release.

**Comment by kencu** [comment]
https://github.com/macports/macports-ports/commit/864aa305471d24ed23dcdfc27fc8b53be239cd56

**Comment by kencu** [comment]
here's hoping it all works out... has a reasonable chance to. If not, we dig in on the individual issues more deeply.

**Comment by kcgen** [comment]
Nice addition, @kencu !

Snow Leopard i686 passed: https://build.macports.org/builders/ports-10.6_i386-builder/builds/94646

Snow Leopard x86-64 and Mojave haven't been started yet; fingers crossed those go through, too!

I've also added fixes to address some more of the warnings.

**Pull request merged**
